File: MemoryReusePool/src/main/java/com/jesen/memoryreusepool/MainActivity.java
package com.jesen.memoryreusepool;

import androidx.appcompat.app.AppCompatActivity;

import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.widget.Button;
import android.widget.ImageView;

import java.io.InputStream;
import java.net.HttpURLConnection;
import java.net.URL;

public class MainActivity extends AppCompatActivity implements Runnable {

    private Button testBtn;
    private ImageView ivTest;

    private final String IMG_URL = "https://tse1-mm.cn.bing.net/th/id/R-C.1496e0b4a2d17b25037cc75ca8328ab0?rik=pGo4LcnY18QrNQ&riu=http%3a%2f%2fpic.lvmama.com%2fuploads%2fpc%2fplace2%2f2016-05-26%2f2732e04c-ffae-49a2-9fbb-72570376cb21.jpg&ehk=x%2baL9ho8oG33wcAqFwq491IpDLG%2flWEETsDegLpn3pI%3d&risl=&pid=ImgRaw&r=0";

    private BitmapReusePool bitmapPool = new ReusePoolImpl(1024*1024*500);

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        testBtn = findViewById(R.id.testBtn);
        ivTest = findViewById(R.id.iv);

        testBtn.setOnClickListener(view -> testRefuse());
    }

    private void testRefuse() {
        new Thread(this).start();

    }

    @Override
    public void run() {
        try {
            URL url = new URL(IMG_URL);
            HttpURLConnection connection = (HttpURLConnection) url.openConnection();
            connection.setConnectTimeout(5000);
            int responseCode = connection.getResponseCode();
            if (HttpURLConnection.HTTP_OK == responseCode) {
                InputStream inputStream = connection.getInputStream();

                // 拿到图片宽和高
                /*BitmapFactory.Options options = new BitmapFactory.Options();
                // 只拿到周围信息，outXXX， outW，outH
                options.inJustDecodeBounds = true;
                BitmapFactory.decodeStream(inputStream, null, options);
                int w = options.outWidth;
                int h = options.outHeight;*/

                int w = 2696;
                int h = 1799;

                BitmapFactory.Options options2 = new BitmapFactory.Options();

                // 拿到复用池  条件： bitmap.isMutable() == true;
                Bitmap bitmapPoolResult = bitmapPool.get(w, h, Bitmap.Config.RGB_565);

                // 如果设置为null，内部就不会去申请新的内存空间，无法复用，依然会重复申请内存，造成：内存抖动，内存碎片
                options2.inBitmap = bitmapPoolResult; // 把复用池的Bitmap 给 inBitmap
                options2.inPreferredConfig = Bitmap.Config.RGB_565; // 2个字节
                options2.inJustDecodeBounds = false;
                options2.inMutable = true; // 符合 复用机制
                final Bitmap bitmap = BitmapFactory.decodeStream(inputStream, null, options2); // 复用内存

                // 添加到复用池
                bitmapPool.put(bitmap);

                new Handler(Looper.getMainLooper()).post(new Runnable() {
                    @Override
                    public void run() {
                        ivTest.setImageBitmap(bitmap);
                    }
                });
            }

        } catch (Exception e) {
            e.printStackTrace();
        }
    }
}

File: bitmaptest/src/main/java/com/jesen/bt/bitmaptest/MainActivity.kt
package com.jesen.bt.bitmaptest

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.BitmapRegionDecoder
import android.graphics.Rect
import android.os.Build
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.util.Log
import androidx.annotation.DrawableRes
import java.io.IOException

class MainActivity : AppCompatActivity() {

    private val TAG = "MainActivity"

    // 可复用的bitmap
    private var reuseBitmap: Bitmap? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setContentView(R.layout.activity_main)

        val op = BitmapFactory.Options()
        op.inMutable = true
        reuseBitmap = BitmapFactory.decodeResource(resources, R.drawable._advert, op)

        /**
         * 查看图片加载所需内存大小
         * */
        val bitmap8888 = BitmapFactory.decodeResource(resources, R.drawable._advert)
        val size = bitmap8888.allocationByteCount
        val size2 = bitmap8888.byteCount
        Log.d(TAG, "8888,  size: $size, size2: $size2")

        /**
         * 图片内存优化：修改Config为RGB_565，设置采样参数
         * */
        val options = BitmapFactory.Options()
        options.inPreferredConfig = Bitmap.Config.RGB_565 // 用565,一个像素两字节，内存减半
        options.inSampleSize = 2 // 宽高维度每隔2个像素一次采样，所需内存进一步减小
        val bitmap565 = BitmapFactory.decodeResource(resources, R.drawable._advert, options)
        val size3 = bitmap565.allocationByteCount
        val size4 = bitmap565.byteCount
        Log.d(TAG, "565  size: $size3, size2: $size4")
    }

    /**
     * Bitmap内存复用
     * */
    fun getBitmap(): Bitmap {
        val options = BitmapFactory.Options()
        options.inJustDecodeBounds = true
        BitmapFactory.decodeResource(resources, R.drawable._advert, options)
        if (canUseForInBitmap(reuseBitmap, options)) {
            options.inMutable = true
            options.inBitmap = reuseBitmap
        }
        options.inJustDecodeBounds = false
        return BitmapFactory.decodeResource(resources, R.drawable._advert, options)
    }

    private fun canUseForInBitmap(
        reuseBitmap: Bitmap?,
        targetOptions: BitmapFactory.Options
    ): Boolean {
        val width = targetOptions.outWidth / Math.max(targetOptions.inSampleSize, 1)
        val height = targetOptions.outHeight / Math.max(targetOptions.inSampleSize, 1)

        val byteCount = width * height * getBytesPerPixl(reuseBitmap?.config)
        return byteCount <= reuseBitmap?.allocationByteCount!!
    }

    private fun getBytesPerPixl(config: Bitmap.Config?): Int {
        return when (config) {
            Bitmap.Config.ALPHA_8 -> 1
            Bitmap.Config.RGB_565, Bitmap.Config.ARGB_4444 -> 2
            else -> 4
        }
    }

    /**
     * 大图加载的基础：
     * BitmapRegionDecoder 图片分片显示
     * */
    private fun getReginImage():Bitmap?{
        try {
            val inputStream = assets.open("test.jpg")
            // 设置显示图片的中心区域
            val regionDecoder = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
                BitmapRegionDecoder.newInstance(inputStream)
            } else {
                BitmapRegionDecoder.newInstance(inputStream,false)
            }
            val options = BitmapFactory.Options()
            val bitmap = regionDecoder?.decodeRegion(Rect(0,0,200,200),options)
            return bitmap
        }catch (e:IOException){
            e.printStackTrace()
        }
        return null
    }
}